File: compiler/nodes/expressions/function.go
package expressions

import (
	"slices"

	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/patterns"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type FunctionExpressionNode struct {
	Inputs []database.Node
	Output database.Node
	Facts  *database.Facts

	inputTemporaries []database.Node
}

func (node *FunctionExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseFunctionExpression(parser *syntax.Parser) (*FunctionExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	inputs, err := ParseFunctionExpressionInputs(parser)
	if err != nil {
		return nil, err
	}

	output, err := ParseExpression(parser)
	if err != nil {
		return nil, err
	}

	return &FunctionExpressionNode{
		Inputs: inputs,
		Output: output,
		Facts:  database.NewFacts(span()),
	}, nil
}

func ParseFunctionExpressionInputs(parser *syntax.Parser) ([]database.Node, *syntax.Error) {
	many, err := syntax.ParseMany(parser, 1, patterns.ParseAtomicPattern, syntax.ParseNothing)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("FunctionOperator", syntax.TokenConfig{
		Commit: "in this function",
	})
	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	inputs := make([]database.Node, 0, len(many))
	for _, item := range many {
		inputs = append(inputs, item.Value)
	}

	return inputs, nil
}

func (node *FunctionExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	visitor.PushScope()

	node.inputTemporaries = make([]database.Node, 0, len(node.Inputs))
	for _, pattern := range node.Inputs {
		node.inputTemporaries = append(node.inputTemporaries, visitor.VisitMatching(pattern))
	}

	visitor.Visit(node.Output)

	visitor.PopScope()

	visitor.Constraint(typecheck.TypeConstraint(node, typecheck.FunctionType(node.Inputs, node.Output)))
}

func (node *FunctionExpressionNode) Codegen(c *codegen.Codegen) error {
	if node.inputTemporaries == nil {
		return c.Error(node)
	}

	span := database.GetSpanFact(node)

	c.WriteString(span, "(async (")
	for _, temporary := range node.inputTemporaries {
		c.WriteNode(span, temporary)
		c.WriteString(span, ", ")
	}
	c.WriteString(span, ") => {")
	c.WriteLine()

	for _, temporary := range patterns.CollectTemporaries(node.Inputs...) {
		if slices.Contains(node.inputTemporaries, temporary) {
			continue
		}

		c.WriteString(span, "var ")
		c.WriteNode(span, temporary)
		c.WriteString(span, ";")
		c.WriteLine()
	}

	for _, pattern := range node.Inputs {
		c.WriteString(span, "if (true")
		if err := c.Write(pattern); err != nil {
			return err
		}
		c.WriteString(span, `) {} else { throw new Error("unreachable"); }`)
		c.WriteLine()
	}

	c.WriteString(span, "return ")
	if err := c.Write(node.Output); err != nil {
		return err
	}
	c.WriteString(span, ";")
	c.WriteLine()
	c.WriteString(span, "})")

	return nil
}

File: compiler/src/nodes/expressions/function.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Node, NodeRef},
    nodes::{parse_atomic_pattern, parse_expression, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Captures, Visit, Visitor},
};

#[derive(Debug)]
pub struct FunctionExpressionNode {
    pub inputs: Vec<NodeRef>,
    pub output: NodeRef,
}

impl Node for FunctionExpressionNode {}

pub fn parse_function_expression(
    parser: &mut Parser<'_>,
) -> Result<FunctionExpressionNode, ParseError> {
    let inputs = parse_function_expression_inputs(parser)?;
    let output = parse_expression(parser)?;

    Ok(FunctionExpressionNode { inputs, output })
}

pub fn parse_function_expression_inputs(
    parser: &mut Parser<'_>,
) -> Result<Vec<NodeRef>, ParseError> {
    let inputs = parser
        .parse_many(1, parse_atomic_pattern, |parser| parser.parse_nothing())?
        .into_iter()
        .map(|(node, _)| node)
        .collect::<Vec<_>>();

    parser.token(TokenKind::FunctionOperator)?;
    parser.commit("in this function");
    parser.consume_line_breaks();

    Ok(inputs)
}

impl Visit for FunctionExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        visitor.push_scope();

        for pattern in &self.inputs {
            visitor.matching(pattern, false, false, |visitor| {
                visitor.current_match().root = Some(pattern.clone());
                visitor.current_match().arm = Some(pattern.clone());
                visitor.visit(pattern);
            });

            visitor.edge(pattern, node, "input");
        }

        visitor.visit(&self.output);
        visitor.edge(&self.output, node, "output");

        visitor.pop_scope();

        visitor.constraint(TypeConstraint::new(
            node.clone(),
            visitor.function_type(self.inputs.iter().cloned(), self.output.clone()),
        ));
    }
}

impl Codegen for FunctionExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let Captures(captures) = ctx.get(node).unwrap_or_default();

        ctx.push_instructions();

        ctx.push_conditions();
        for pattern in &self.inputs {
            ctx.codegen(pattern)?;
        }
        let conditions = ctx.pop_conditions();

        ctx.instruction(ir::Instruction::If {
            node: None,
            branches: vec![(conditions, Vec::new(), None)],
            else_branch: None,
        });

        ctx.codegen(&self.output)?;

        ctx.instruction(ir::Instruction::Return {
            value: self.output.clone(),
        });

        let instructions = ctx.pop_instructions();

        ctx.instruction(ir::Instruction::Value {
            node: node.clone(),
            value: ir::Value::Function {
                inputs: self.inputs.clone(),
                captures: Vec::from_iter(captures),
                instructions,
            },
        });

        Ok(())
    }
}
